### ISSUE TITLE
v17 - documentation

### ISSUE THREAD

**Issue by gr2m** [octokit/octokit.js#1595]
This issue is part of #1546. 

In terms of documentation, everything is up for discussion. Overarching goals are

1. Simple setup
2. Low barrier for documentation contributors
3. Minimal maintenance overhead

## Current state (v16)

- Current website: https://octokit.github.io/rest.js/
- The website is built with [Gatsby](https://www.gatsbyjs.org/), the source code lives at https://github.com/octokit/rest.js/tree/master/docs
- I've created a custom plugin to load nodes for all the REST API endpoint methods, which is probably 95% of the documentation: https://github.com/octokit/rest.js/blob/master/docs/plugins/gatsby-source-octokit-routes/gatsby-node.js
  - The endpoint definitions are coming from https://github.com/octokit/routes. It scrapes [GitHub's REST API docs](https://developer.github.com/v3) and turns it into a JSON API spec
  - I've created a GraphQL API that uses the OpenAP as source, for simpler querying of the endpoints: https://octokit-routes-graphql-server.now.sh/
  - The GraphQL endpoint is used to generate the code for the endpoint methods, the TypeScript definitions, and the docs: https://github.com/octokit/rest.js/tree/master/scripts/update-endpoints
- The website is continuously build and deployed from master

## What is changing (v17)

- `@octokit/rest` v17 will be [roughly 10 lines of code when it's done](https://github.com/octokit/rest.js/blob/beta/src/index.ts). It will be built upon a plugin architecture that I've built out over the past year.

  ```
  @octokit/rest
  ├─┬ @octokit/core@2.4.0
  │ ├── @octokit/auth-token@2.4.0
  │ ├── @octokit/graphql@4.3.1
  │ └─┬ @octokit/request@5.3.1
  │   ├── @octokit/endpoint@5.5.2
  │   └── @octokit/request-error@1.2.1
  ├── @octokit/plugin-paginate-rest@2.0.1
  ├── @octokit/plugin-request-log@1.0.0
  └── @octokit/plugin-rest-endpoint-methods@3.0.0
  ```

- All the REST API endpoint methods are now living in [`@octokit/plugin-rest-endpoint-methods`](https://github.com/octokit/plugin-rest-endpoint-methods.js). If possible, I would avoid loading the endpoints definitions again in `@octokit/rest`, to prevent that the `@octokit/rest` documentation gets out of sync with its code.

### Must

- We will continue to maintain v16 for some time. The documentation must remain accessible.

### Should

- When opening https://octokit.github.io/rest.js/, the browser should redirect to the latest version, e.g. `https://octokit.github.io/rest.js/v17.2.3`
- We only need to deploy the latest version for each breaking version (`17.x`, `16.x`, etc) of the documentation
- Deployment of docs for every pull request (ideally PRs including changes to documentation only)

### Could

- Documentation could be distributed across plugins, just like the code. When combining plugins into a library such as `@octokit/rest`, the documentation sources from the plugins could be combined into one. For example, https://github.com/octokit/plugin-rest-endpoint-methods.js could generate Markdown files for every single endpoint, https://github.com/octokit/plugin-paginate-rest.js could have a Markdown file for a "Pagination" section.

**Comment by trevorblades** [comment]
I could help out with this versioned docs challenge! I would recommend leveraging `gatsby-source-git` to source content from a separate git branch. This gives the old docs a place to live outside of the current-version codebase.

At Apollo, I also experimented with generating docs based on the state of the git repo at a specific version tag using `simple-git` in my Gatsby `createPages` hook. This was convenient, but didn't easily allow for docs updates without publishing a new version (and tag) for older versions. We ended up using version branches because it gave us this extra flexibility.

This work would involve the following tasks:

1. Create a `version-16` (or other name) branch based on `master`
2. In the `beta` branch, install `gatsby-source-git` and configure it to source markdown files from the `docs/src/pages/api` directory on the `version-16` branch
3. Create a page template based on `src/pages/index.js` that can be query for content filtered by version
4. Loop through branches sourced from git in the `createPages` Gatsby API and [programatically create version pages](https://www.gatsbyjs.org/tutorial/part-seven/#creating-pages) for each using the template component

Let me know what you think! I can put together a PR that demonstrates this approach 👍
